feat(docs): fetch remote docs from GitHub

CodyJasonBennett opened the pull request:
Implements #86 by fetching docs from libraries' respective repositories instead of solely from this repo.

Some things to address:
 - [x] target configuration (opt-in via `data/libraries` under the `docs` key, configured to fetch from this repo by default)
 - [x] file-based routing with categories (read as `<category-name>/<page-name>.md(x)` and will defer to using MDX frontmatter for page links and titles)
 - [x] automatic rewrites (no more redirect pages, external rewrites are handled in next.config for Jotai)
 - [x] fetch configured targets via git
 - [x] wiki support
 - [x] vanilla frontmatter support (commented out, but the same syntax as MDX)
 - [x] vanilla markdown support (we rely on heavy MDX syntax, but some repos don't use it -- maybe magically insert some components)
 - [x] update strategy (ISR builds don't work at run-time w/o git and local storage, will have to configure git/deploy hooks)
 - ~~streamline syntax (if we do support MDX/vanilla magic, we have a bunch of redundant components like Demo, Sandbox, CodeSandbox)~~ -- we'll have to leave this to authors, especially since they can ship custom components in their repos

dai-shi commented on pages/[...slug].tsx:
```diff
@@ -53,14 +51,88 @@ export default function PostPage({ toc, source, allDocs, nav, frontMatter }) {
+  const { repo, dir, tag = 'master' } = docs
+
+  const { tree } = await fetch(
+    `https://api.github.com/repos/${repo}/git/trees/${tag}?recursive=1`
```
curious if it's technically possible fetch from github wiki pages.
ex. https://github.com/pmndrs/valtio.wiki.git

CodyJasonBennett replied:
I'll have to give that a try. Cloning the repository (wikis can be cloned as repos) is a more robust way of doing this, although slower.

We'd have to get creative in caching by SHA or versioning and busting cache to trigger rebuilds.

CodyJasonBennett commented:
Vercel doesn't like the formatting but this is what Valtio's wiki looks like.

![image](https://user-images.githubusercontent.com/23324155/142747050-eef003b9-b727-4fca-b3eb-7a08d310eaf3.png)

dai-shi commented:
wow, seems promising. having metadata in comments is totally acceptable.

CodyJasonBennett commented:
Discovered some issues/pitfalls with the current system for markdown:
 - comments are left in the final eval/JSX code, so they have to be removed
 - vanilla markdown is evaluated as MDX, so stuff like `import ...` will break
 - escaped MDX syntax is evaluated as though it isn't escaped (\`jsx here\`)

dai-shi commented:
Is there a workaround on wiki side? Not sure if I get what "isolated there" means.

CodyJasonBennett commented:
I'll take a look. The issues are bugs/unsupported features with our current dependencies.

CodyJasonBennett commented:
It seems that the use of space and `<br />` around paragraphs is tripping up our markdown parsers, so they're not parsed into components but plain text. Combined with eval, that will have the paragraph's contents evaluated as JSX.

We can possibly clean this up within the Wiki.

As-is from wiki:
![image](https://user-images.githubusercontent.com/23324155/142780234-129a1faf-5816-4ff8-a957-08b88c1a3145.png)

Tidied:
![image](https://user-images.githubusercontent.com/23324155/142780224-e518378d-4b6c-455e-95b2-942913878142.png)

CodyJasonBennett commented on pages/valtio/index.ts:
```diff
@@ -0,0 +1,3 @@
+import redirect from '../../utils/redirect'
+
+export default redirect('/valtio/home')
```
I think we can do away with this by redirecting paths to their first child, respecting `nav` order.

CodyJasonBennett commented:
I'm hosting the wiki under [this repo](https://github.com/CodyJasonBennett/valtio), but you can see it deployed under [/valtio](https://website-6nhkftsfw-pmndrs.vercel.app/valtio/home) now.

I have frontmatter configured to preserve the nav order found on the wiki, and I've cleaned them up by replacing `<br />` with new lines and replaced h1s with h2s (`#` => `##`).

We might be able to automatically streamline docs with those changes.

CodyJasonBennett commented on pages/api/github/push.ts:
```diff
@@ -0,0 +1,41 @@
+import { MARKDOWN_REGEX } from 'utils/mdx'
+import { getDocs } from 'utils/docs'
+import { data } from 'data/libraries'
+
+interface GithubWebhook extends NextApiRequest {
+  body: {
+    repository: {
+      full_name: string
+    }
+    ref: string
+    head_commit: {
+      added: string[]
+      removed: string[]
+      changed: string[]
+    }
+  }
+}
+
+export default async function handler(req: GithubWebhook, res: NextApiResponse) {
+  const { repository, ref, head_commit } = req.body
+
+  // Check for repo
+  const repo = data.find((lib) => lib.docs.repo === repository.full_name)
+  if (!repo) return res.status(400).json({ error: 'Invalid repository' })
+
+  // Check if on docs branch
+  if (ref !== `ref/heads/${repo.docs.branch}`) return res.status(304)
+
+  // Check if docs were modified
+  const files = [...head_commit.added, head_commit.removed, head_commit.changed]
+  const modified = files.filter(
+    (file: string) => file.startsWith(repo.docs.dir) && MARKDOWN_REGEX.test(file)
+  )
+  if (!modified.length) return res.status(304)
+
+  // Update docs
+  await getDocs(repo.id, true)
+
+  return res.status(201).json({ message: `${modified.length} docs updated` })
+}
```
Maybe this isn't necessary if we can pull and check for changes via git. We can leverage [Next's ISR](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration) this way to rebuild on remote changes.

CodyJasonBennett commented on utils/mdx.ts:
```diff
@@ -0,0 +1,113 @@
+  // Update repo if on disk, otherwise clone it
+  const repoDir = path.resolve(process.cwd(), `temp/${repo}-${branch}`)
+  if (!fs.existsSync(repoDir)) clone({ repo, dir, branch })
+  // else execSync(`git -C ${repoDir} pull`)
```
Need to figure out how to pull changes while not authenticated (read-only).

CodyJasonBennett commented on pages/[...slug].tsx:
```diff
@@ -54,21 +51,42 @@ export default function PostPage({ toc, source, allDocs, nav, frontMatter }) {
-  const postData = fs.readFileSync(postFilePath)
-  const { content, data } = matter(postData)
+  const [lib] = params.slug
+  const docs = await getDocs(lib, false)
```
Leaving the invalidate param as `false` for now until we can find a workaround.

With invalidation enabled:
![image](https://user-images.githubusercontent.com/23324155/143611900-9cef40f7-5e83-467e-b76a-4230f9230f21.png)

CodyJasonBennett commented:
Just noticed that our build times are kinda slow on Vercel.

<img src="https://user-images.githubusercontent.com/23324155/143659688-0a857fd7-21c6-441a-ba84-1dedbcddefb4.png" width="500px" />


Is there any throttling I should know about? Not getting this on my machines.

CodyJasonBennett commented on pages/[...slug].tsx:
```diff
@@ -50,27 +50,35 @@ export default function PostPage({ allDocs, nav, toc, data, source }: PostPagePr
   )
 }
 
+interface PostPageProps extends Partial<PostProps> {
+  redirect?: string
+}
+
+export default function PostPage({ redirect, ...rest }: PostPageProps) {
+  const router = useRouter()
+
+  useLayoutEffect(() => {
+    if (redirect) return void router.push(redirect)
+  }, [redirect, router])
+
+  return redirect ? null : <Post {...rest} />
+}
+
```
Maybe we can serve the page specified as `redirect` instead? Ideally, we can change the URL as well without consequence. Looking into [*shallow routing*](https://nextjs.org/docs/routing/shallow-routing).

CodyJasonBennett commented on pages/api/update.ts:
```diff
@@ -0,0 +1,63 @@
+    if (!changes.length) return res.status(304)
+
+    // Update docs
+    await fetch(process.env.VERCEL_DEPLOY_WEBHOOK, { method: 'POST' })
```
The webhook is currently aimed at this branch. Will need to point it to docs (prod) once merged.
